### PR TITLE
Complex topic whitelisting from file

### PR DESCRIPTION
**Pull request by VivekPipaliya23**
## Proposed Changes
Currently the complex topic is added via the vmq-admin CLI and it is stored in redis.
This PR adds complex topics using a file and updates the complex topic trie in give interval.
Removes the implementations of storing topics in redis.

## Types of Changes

- [ ] Bugfix (non-breaking change which fixes issue #XXXX)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

## Checklist

- [x] I have read the `CODE_OF_CONDUCT.md` document
- [x] All tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if needed)
- [ ] Any dependent changes have been merged and published in related repositories
- [ ] I have updated changelog (At the bottom of the release version)
- [ ] I have squashed all my commits into one before merging

### CODE REVIEW

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_enhanced_auth/src/vmq_enhanced_auth.erl [hunk 1]
```diff
@@ -531,7 +531,7 @@ age_entries(T) ->
 del_aged_entries() ->
     lists:foreach(fun del_aged_entries/1, ?TABLES).
 del_aged_entries(T) ->
-    ets:match_delete(T, {'_', 2}).
+    ets:match_delete(T, {'_', 2, '_'}).
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_enhanced_auth/src/vmq_enhanced_auth.erl:534]: we are adding label at the end as third argument.
`t(read, token, Topic, Label) -> {vmq_enhanced_auth_acl_read_token, {Topic, 1, Label}};`
so at the time of `del_aged_entries` it needs to match third argument.

File: apps/vmq_server/priv/vmq_server.schema [hunk 2]
```diff
@@ -1771,3 +1771,16 @@
          true;
     (Name) -> vmq_schema_util:file_is_readable(Name)
  end}.
+
+%% @doc Set the path to an complex access control list file.
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/priv/vmq_server.schema]: Updated

File: apps/vmq_server/priv/vmq_server.schema [hunk 3]
```diff
@@ -1771,3 +1771,16 @@
...
+{mapping, "vmq_trie.acl_file", "vmq_server.file", [
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/priv/vmq_server.schema]: Updated to `complex_trie_file`

File: apps/vmq_server/priv/vmq_server.schema [hunk 4]
```diff
@@ -1771,3 +1771,16 @@
...
+                                               {default, "{{platform_etc_dir}}/vmq.trie"},
+                                               {datatype, file}
+                                              ]}.
+
+%% @doc set the acl reload interval in seconds, the value 0 disables
+%% the automatic reloading of the complex acl file.
+{mapping, "vmq_trie.acl_reload_interval", "vmq_server.interval", [
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/priv/vmq_server.schema]: Updated to `complex_trie_reload_interval`

File: apps/vmq_server/priv/vmq_server.schema [hunk 5]
```diff
@@ -1771,3 +1771,16 @@
...
     (Name) -> vmq_schema_util:file_is_readable(Name)
  end}.
+
+%% @doc Set the path to an complex access control list file.
+{mapping, "vmq_trie.acl_file", "vmq_server.file", [
+                                               {default, "{{platform_etc_dir}}/vmq.trie"},
+                                               {datatype, file}
+                                              ]}.
+
+%% @doc set the acl reload interval in seconds, the value 0 disables
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/priv/vmq_server.schema]: Updated

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 6]
```diff
@@ -134,20 +131,44 @@ fold_local_shared_subscriber_info(
...
-            ),
-            [[?SADD, "wildcard_topics", term_to_binary(T)] | Acc]
+load_from_file(File) ->
+    case file:read_file(File) of
+        {ok, BinaryData} ->
+            FileContent = binary_to_list(BinaryData),
+            Lines = string:tokens(FileContent, "\n"),
+            Topics = create_list(Lines),
+            TopicList = parse_topic_list(Topics),
+            update_complex_topics(TopicList);
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl]: Updated the fn name

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 7]
```diff
@@ -134,20 +131,44 @@ fold_local_shared_subscriber_info(
...
-            lists:foreach(
-                fun(Node) -> safe_rpc(Node, ?MODULE, add_complex_topic, ["", T]) end, Nodes
-            ),
-            [[?SADD, "wildcard_topics", term_to_binary(T)] | Acc]
+load_from_file(File) ->
+    case file:read_file(File) of
+        {ok, BinaryData} ->
+            FileContent = binary_to_list(BinaryData),
+            Lines = string:tokens(FileContent, "\n"),
+            Topics = create_list(Lines),
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl]: Removed the create_list fn.

File: apps/vmq_server/src/vmq_server_cli.erl [hunk 8]
```diff
@@ -73,8 +73,6 @@ register_cli() ->
 
     vmq_tracer_cli:register_cli(),
 
-    vmq_reg_redis_trie_add_wildcard_topics_cmd(),
-    vmq_reg_redis_trie_remove_wildcard_topics_cmd(),
     vmq_reg_redis_trie_show_wildcard_topics_cmd(),
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_server_cli.erl]: Updated to `vmq_trie_show_complex_topics_cmd`

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 9]
```diff
@@ -134,20 +131,44 @@ fold_local_shared_subscriber_info(
...
 
-add_complex_topics(Topics) ->
-    Nodes = vmq_cluster_mon:nodes(),
-    Query = lists:foldl(
-        fun(T, Acc) ->
-            lists:foreach(
-                fun(Node) -> safe_rpc(Node, ?MODULE, add_complex_topic, ["", T]) end, Nodes
-            ),
-            [[?SADD, "wildcard_topics", term_to_binary(T)] | Acc]
+load_from_file(File) ->
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl:134]: `parse_topic_list` should only return valid topics

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 10]
```diff
@@ -134,20 +131,44 @@ fold_local_shared_subscriber_info(
...
+    case file:read_file(File) of
+        {ok, BinaryData} ->
+            FileContent = binary_to_list(BinaryData),
+            Lines = string:tokens(FileContent, "\n"),
+            Topics = create_list(Lines),
+            TopicList = parse_topic_list(Topics),
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl]: Updated `parse_topic_list` to return list of only valid topics.

[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl]: Added spec for `parse_topic_list`.

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 11]
```diff
@@ -427,6 +417,32 @@ trie_delete_path(MP, [{Node, Word, _} | RestPath]) ->
...
 
+parse_topic_list(Topics) ->
+    lists:foldl(
+        fun(T, Acc) ->
+            Topic = list_to_binary(string:strip(T)),
+            case vmq_topic:validate_topic(subscribe, Topic) of
+                {ok, ParsedTopic} ->
+                    case vmq_topic:contains_wildcard(ParsedTopic) of
+                        true -> [ParsedTopic | Acc];
+                        false -> Acc
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl]: Added log

File: apps/vmq_server/src/vmq_reg_redis_trie.erl [hunk 12]
```diff
@@ -427,6 +417,32 @@ trie_delete_path(MP, [{Node, Word, _} | RestPath]) ->
...
+                    end;
+                _ ->
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_reg_redis_trie.erl:432]: Added log

**Review by dhruvjain99 — APPROVED** [review]

**Pull request merged**